README.md
# saisoku - Fast file transfer orchestration pipeline

<img align="left" width="226" height="200" src="docs/saisoku.png?raw=true" hspace="5" vspace="5" alt="saisoku">

Saisoku is a Python (2.7, 3.6 tested) package that helps you build complex pipelines of batch file/directory transfer/sync jobs. It supports threaded transferring of files locally, over network mounts, or HTTP. With Saisoku you can also transfer files to and from AWS S3 buckets and sync directories using Rclone and keep directories in sync "real-time" with Watchdog.

Saisoku includes a Transfer Server and Client which support copying over TCP sockets.

Saisoku uses Luigi for task management and web ui. To learn more about Luigi, see it's [github](https://github.com/spotify/luigi) or [readthedocs](https://luigi.readthedocs.io/en/stable/index.html).


[![License](https://img.shields.io/github/license/shirosaidev/saisoku.svg?label=License&maxAge=86400)](./LICENSE)
[![Release](https://img.shields.io/github/release/shirosaidev/saisoku.svg?label=Release&maxAge=60)](https://github.com/shirosaidev/saisoku/releases/latest)
[![Sponsor Patreon](https://img.shields.io/badge/Sponsor%20%24-Patreon-brightgreen.svg)](https://www.patreon.com/shirosaidev)
[![Donate PayPal](https://img.shields.io/badge/Donate%20%24-PayPal-brightgreen.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=CLF223XAS4W72)


## Requirements
- luigi
- tornado
- scandir
- pyfastcopy
- tqdm
- requests
- beautifulsoup4
- boto3
- watchdog

Install above python modules using pip

```sh
$ pip install -r requirements.txt
```

## Download

```shell
$ git clone https://github.com/shirosaidev/saisoku.git
$ cd saisoku
```
[Download latest version](https://github.com/shirosaidev/saisoku/releases/latest)


## How to use

## Start Luigi

Create directory for state file for Luigi
```sh
$ mkdir /usr/local/var/luigi-server
```
Start Luigi scheduler daemon in foreground with
```sh
$ luigid --state-path=/usr/local/var/luigi-server/state.pickle
```
or in the background with
```sh
$ luigid --background --state-path=/usr/local/var/luigi-server/state.pickle --logdir=/usr/local/var/log
```
It will default to port 8082, so you can point your browser to http://localhost:8082 to access the web ui.

## Configure Boto 3

If you are going to use the S3 copy Luigi tasks, first start be setting up Boto 3 (aws sdk python module) with the quick start instructions at [boto 3 github](https://github.com/boto/boto3).

## Usage - Luigi tasks

### Local/network mount copy

With the Luigi centralized scheduler running, we can send a copy files task to Luigi 
```sh
$ python run_luigi.py CopyFiles --src /source/path --dst /dest/path
```

See below for the different [parameters](#using-saisoku-module-in-python) for each Luigi task.

### Tarball package copy
To run a copy package task, which will create a tar.gz (gzipped tarball) file containing all files at src and copy the tar.gz to dst
```sh
$ python run_luigi.py CopyFilesPackage --src /source/path --dst /dest/path
```

### HTTP copy

Start up 2 Saisoku http servers, the get requests from saisoku clients will be load balanced across these.
```sh
$ python saisoku_server.py --httpserver -p 5005 -d /src/dir
$ python saisoku_server.py --httpserver -p 5006 -d /src/dir
```
This will create an index.html file on http://localhost:5005 serving up the files in /src/dir.

To send a HTTP copy files task to Luigi
```sh
$ python run_luigi.py CopyFilesHTTP --src http://localhost --dst /dest/path --ports [5005,5006] --threads 2
```

### S3 copy

To copy a local file to s3 bucket
```sh
$ python run_luigi.py CopyLocalFileToS3 --src /source/file --dst s3://bucket/foo/bar
```

s3 bucket object to local file
```sh
$ python run_luigi.py CopyS3lFileToLocal --src s3://bucket/foo/bar --dst /dest/file
```

### Rclone sync

Saisoku can use Rclone to sync directories, etc. First, make sure you have [Rclone](https://rclone.org/) installed and in your PATH.

To to do a dry-run sync from source to dest using Rclone:
```sh
$ python run_luigi.py SyncDirsRclone --src /source/path --dst /dest/path
```

To sync from source to dest using Rclone
```sh
$ python run_luigi.py SyncDirsRclone --src /source/path --dst /dest/path --cmdargs '["-vv"]'
```

To change the subcommand that Rclone uses (default is sync)
```sh
$ python run_luigi.py SyncDirsRclone --src /source/path --dst /dest/path --command 'subcommand'
```

### Watchdog directory sync

Saisoku can use watchdog to keep directories synced in "real-time". First, make sure you have rsync installed and in your PATH.

To keep directories in sync from source to dest using Watchdog
```sh
$ python run_luigi.py SyncDirsWatchdog --src /source/path --dst /dest/path
```

## Usage - Server -> Client transfer

Start up Saisoku Transfer server listening on all interfaces on port 5005 (default)
```sh
$ python saisoku_server.py --host 0.0.0.0 -p 5005
```
Run client to download file from server
```sh
$ python saisoku_client.py --host 192.168.2.3 -p 5005 /path/to/file
```


## Log file

Saisoku output get logged to os env TEMP/TMPDIR directory in `saisoku.log` file.


## Using saisoku module in Python

### ThreadedCopy

Saisoku's `ThreadedCopy` class requires two parameters:

`src` source directory containing files you want to copy

`dst` destination directory of where you want the files to go (directory will be created if not there already)

Optional parameters:

`filelist` optional txt file containing one filename per line of files in src directory (not full path)

`ignore` optional ignore files list, example `['*.pyc', 'tmp*']`

`threads` number of worker copy threads (default 16)

`symlinks` copy symlinks (default False)

`copymeta` copy file stat info (default True)


```
>>> from saisoku import ThreadedCopy

>>> ThreadedCopy(src='/source/dir', dst='/dest/dir', filelist='filelist.txt')
calculating total file size..
100%|██████████████████████████████████████████████████████████| 173/173 [00:00<00:00, 54146.30files/s]
copying 173 files..
100%|██████████████████████████████████████████████| 552M/552M [00:06<00:00, 97.6MB/s, file=dk-9.4.zip]
```

### ThreadedHTTPCopy

Saisoku's `ThreadedHTTPCopy` class requires two parameters:

`src` source http tornado server (tserv) serving a directory of files you want to copy

`dst` destination directory of where you want the files to go (directory will be created if not there already)

Optional parameters:

`threads` number of worker copy threads (default 1)

`ports` tornado server (tserv) ports, these ports will be load balanced (default [5000])

`fetchmode` file get mode, either requests or urlretrieve (default urlretrieve)

`chunksize` chunk size for requests fetchmode (default 8192)

```
>>> from saisoku import ThreadedHTTPCopy

>>> ThreadedHTTPCopy('http://localhost', '/dest/dir')
```

### Rclone

Saisoku's `Rclone` class requires two parameters:

`src` source directory of files you want to sync

`dst` destination directory of where you want the files to go

Optional parameters:

 def __init__(self, src, dst, flags=[], command='sync', cmdargs=[]):

`flags` a list of Rclone flags (default [])

`command` subcommand you want Rclone to use (default sync)

`cmdargs` a list of command args to use (default ['--dry-run', '-vv'])

```
>>> from saisoku import Rclone

>>> Rclone('/src/dir', '/dest/dir')
```

### Watchdog

Saisoku's `Watchdog` class requires two parameters:

`src` source directory of files you want to sync

`dst` destination directory of where you want the files to go

Optional parameters:

 def __init__(self, src, dst, recursive, patterns, ignore_patterns, ignore_directories, case_sensitive)

`recursive` bool used for recurisvely checking all sub directories for changes (default True)

`patterns` file name patterns to use when checking for changes (default *)

`ignore_patterns` file name patterns to ingore when checking for changes (default *)

`ignore_directories` bool used for ignoring directories (default False)

`case_sensitive` bool used for being case sensitive (default True)


```
>>> from saisoku import Watchdog

>>> Watchdog('/src/dir', '/dest/dir')
```

## Patreon
If you are a fan of the project or using Saisoku in production, please consider becoming a [Patron](https://www.patreon.com/shirosaidev) to help advance the project.


CHANGELOG.MD
# saisoku Change Log

## [0.1-b.5] = 2020-09-11
### added
- watchdog task (SyncDirsWatchdog class) for keeping two directories (optionally recursive) in sync in "real-time" (requires rsync in PATH)
- requirement for watchdog py module

## [0.1-b.4] = 2019-07-11
### fixed
- TypeError issue with CopyFilesHTTP and python3

## [0.1-b.3] = 2019-02-22
### added
- copy package task (CopyFilesPackage class) for creating tar.gz of files in src directory and copying it to dst
- saisoku transfer server and client
- luigi tasks for copying files to and from aws s3 buckets
- requirement for boto3 (aws sdk python module)
- rclone Luigi task, requires rclone installed and in PATH
- Rclone class to saisoku module
### changed
- improved tserv.py
### fixed
- encoding issue with non-ascii characters and http file copy

## [0.1-b.2] = 2019-02-20
### added
- http file copying
- ThreadedHTTPCopy class
- additional requirements in requirements.txt for http file copying (tornado, requests, beautifulsoup4)

## [0.1-b.1] = 2019-02-19
### note
- first release


tserv.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
"""tsver.py
Saisoku is a Python module that helps you build complex pipelines of batch file copying jobs.

See README.md or https://github.com/shirosaidev/saisoku
for more information.

Author: shirosai <cpark16@gmail.com>

Copyright (C) Chris Park 2019
saisoku is released under the Apache 2.0 license. See
LICENSE for the full license text.
"""

import os
import sys
from argparse import ArgumentParser
from scandir import scandir
import logging
import signal

import tornado.ioloop
import tornado.web
import tornado.template
import tornado.httpserver


logging.basicConfig(level=logging.DEBUG)


class IndexHandler(tornado.web.RequestHandler):
    SUPPORTED_METHODS = ['GET']

    def get(self, path):
        """ GET method to list contents of directory or
        write index page if index.html exists."""

        # remove heading slash
        path = path[1:]

        for index in ['index.html', 'index.htm']:
            index = os.path.join(path, index)
            if os.path.exists(index):
                with open(index, 'rb') as f:
                    self.write(f.read())
                    self.finish()
                    return
        html = self.generate_index(path)
        self.write(html)
        self.finish()

    def generate_index(self, path):
        """ generate index html page, list all files and dirs.
        """
        if path:
            files = [item for item in scandir(path)]
        else:
            files = [item for item in scandir('.')]
        files = [(item.name + '/', 0) if item.is_dir() else (item.name, str(item.stat().st_size)) for item in files]
        html_template = """
        <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 3.2 Final//EN"><html>
        <title>Directory listing for /{{ path }}</title>
        <body>
        <h2>Directory listing for /{{ path }}</h2>
        <hr>
        <ul>
        {% for filename, size in files %}
        <li><a title="{{ size }}" href="{{ filename }}">{{ filename }}</a>
        {% end %}
        </ul>
        <hr>
        </body>
        </html>
        """
        t = tornado.template.Template(html_template)
        return t.generate(files=files, path=path)


class StaticFileHandler(tornado.web.StaticFileHandler):
    
    def write(self, chunk):
        super(StaticFileHandler, self).write(chunk)
        #logging.debug('write called')
        self.flush()


def stop_server(signum, frame):
    tornado.ioloop.IOLoop.instance().stop()
    logging.info('Stopped!')


def start_http_server(args):
    os.chdir(args.dir)
    logging.debug('cwd: %s' % args.dir)
    try:
        application = tornado.web.Application([
        (r'(.*)/$', IndexHandler,),
        (r'/(.*)$', StaticFileHandler, {'path': args.dir}),
        ])
        signal.signal(signal.SIGINT, stop_server)

        http_server = tornado.httpserver.HTTPServer(application)
        http_server.listen(args.port)
        logging.info('Starting HTTP server on 0.0.0.0 port %d ...' % args.port)
        tornado.ioloop.IOLoop.instance().start()
    except KeyboardInterrupt:
        signal.signal(signal.SIGINT, stop_server)


run_luigi.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""run_luigi.py
Saisoku is a Python module that helps you build complex pipelines of batch file copying jobs.

See README.md or https://github.com/shirosaidev/saisoku
for more information.

Author: shirosai <cpark16@gmail.com>

Copyright (C) Chris Park 2019-2020
saisoku is released under the Apache 2.0 license. See
LICENSE for the full license text.
"""

import luigi
from luigi.contrib.s3 import S3Target, S3Client

import os
import time
import logging

from saisoku import logger


class CopyFiles(luigi.Task):
    src = luigi.Parameter()
    dst = luigi.Parameter()
    threads = luigi.IntParameter(default=16)
    filelist = luigi.OptionalParameter(default=None)
    symlinks = luigi.BoolParameter(default=False)
    ignore = luigi.OptionalParameter(default=None)
    copymeta = luigi.BoolParameter(default=True)

    #def output(self):
    #    return []

    #def requires(self):
    #    return []

    def run(self):
        from saisoku import ThreadedCopy
        
        ThreadedCopy(src=self.src, dst=self.dst, threads=self.threads, filelist=self.filelist, 
                    symlinks=self.symlinks, ignore=self.ignore, copymeta=self.copymeta)


class CopyFilesHTTP(luigi.Task):
    src = luigi.Parameter()
    dst = luigi.Parameter()
    threads = luigi.IntParameter(default=1)
    ports = luigi.ListParameter(default=[5005])
    fetchmode = luigi.Parameter(default='urlretrieve')
    chunksize = luigi.IntParameter(default=8192)

    def run(self):
        from saisoku import ThreadedHTTPCopy

        ThreadedHTTPCopy(src=self.src, dst=self.dst, threads=self.threads, ports=self.ports, 
                        fetchmode=self.fetchmode, chunksize=self.chunksize)


class PackageDirectory(luigi.Task):
    src = luigi.Parameter()
    filelist = luigi.Parameter(default='saisoku_filelist.txt')

    def output(self):
        return luigi.LocalTarget(self.filelist)

    def run(self):
        import tarfile
        from scandir import scandir
        import os

        archives = []
        src_path = os.path.abspath(self.src)
        # create tar.gz of all files in directory
        archive_name = '{}_saisoku_archive.tar.gz'.format(os.path.basename(src_path))
        logger.info('Compressing files to %s...' % archive_name)
        tar = tarfile.open(archive_name, "w:gz")
        for item in scandir(self.src):
            if item.is_file():
                logger.debug('  Adding %s...' % item.name)
                tar.add(item.path, item.name)
        tar.close()
        archives.append(archive_name)

        with self.output().open('w') as f:
            for archive in archives:
                f.write('{archive}\n'.format(archive=archive))


class CopyFilesPackage(luigi.Task):
    src = luigi.Parameter()
    dst = luigi.Parameter()
    threads = luigi.IntParameter(default=16)
    filelist = luigi.Parameter(default='saisoku_filelist.txt')
    symlinks = luigi.BoolParameter(default=False)
    ignore = luigi.OptionalParameter(default=None)
    copymeta = luigi.BoolParameter(default=True)
    cleanup = luigi.BoolParameter(default=True)

    def output(self):
        return luigi.LocalTarget(self.filelist)

    def requires(self):
        return [PackageDirectory(src=self.src, filelist=self.filelist)]

    def run(self):
        from saisoku import ThreadedCopy
        import os

        ThreadedCopy(src=self.src, dst=self.dst, threads=self.threads, filelist=self.filelist, 
                symlinks=self.symlinks, ignore=self.ignore, copymeta=self.copymeta, package=True)

        if self.cleanup:
            logger.info('Cleaning up %s...' % self.src)
            with self.input()[0].open() as f:
                filename = f.read().rstrip('\n')
                logger.debug('  Removing %s...' % filename)
                os.remove(filename)  # delete tar.gz file in src
            logger.debug('  Removing %s...' % self.filelist)
            os.remove(self.filelist)  # delete file list


class S3File(luigi.ExternalTask):
    src = luigi.Parameter()

    def output(self):
        return S3Target(self.src)


class CopyS3FileToLocal(luigi.Task):
    src = luigi.Parameter()
    dst = luigi.Parameter()

    def output(self):
        return luigi.LocalTarget(self.dst)

    def requires(self):
        return S3File(src=self.src)

    def run(self):
        """
        examples:
        * s3://bucket/foo/bar.txt
        * s3://bucket/foo/bar.txt?aws_access_key_id=xxx&aws_secret_access_key=yyy
        """
        def copy():
            with self.input().open('r') as infile:
                for line in infile:
                    yield line

        logger.info('Copying %s to %s ...' % (self.src, self.dst))
        t = time.time()
        with self.output().open('w') as outfile:
            for line in copy():
                outfile.write(line)
        t = round(time.time() - t, 2)
        size = round(os.path.getsize(self.dst) / 1024 * 1.0, 2)  # in KB
        kb_per_sec = round(size / t, 2)
        logger.info('Done. Copied %s KB in %s seconds (%s KB/s)' % (size, t, kb_per_sec))


class LocalFile(luigi.ExternalTask):
    src = luigi.Parameter()

    def output(self):
        return luigi.LocalTarget(self.src)


class CopyLocalFileToS3(luigi.Task):
    src = luigi.Parameter()
    dst = luigi.Parameter()

    def output(self):
        return S3Target(self.dst)

    def requires(self):
        return LocalFile(src=self.src)

    def run(self):
        def copy():
            with self.input().open('r') as infile:
                for line in infile:
                    yield line

        logger.info('Copying %s to %s ...' % (self.src, self.dst))
        t = time.time()
        with self.output().open('w') as outfile:
            for line in copy():
                outfile.write(line)
        t = round(time.time() - t, 2)
        size = round(os.path.getsize(self.src) / 1024 * 1.0, 2)  # in KB
        kb_per_sec = round(size / t, 2)
        logger.info('Done. Copied %s KB in %s seconds (%s KB/s)' % (size, t, kb_per_sec))


class SyncDirsRclone(luigi.Task):
    src = luigi.Parameter()
    dst = luigi.Parameter()
    flags = luigi.ListParameter(default=[])
    command = luigi.Parameter(default='sync')
    cmdargs = luigi.ListParameter(default=['--dry-run', '-vv'])

    def run(self):
        from saisoku import Rclone
        
        Rclone(src=self.src, dst=self.dst, flags=self.flags, command=self.command, cmdargs=self.cmdargs)


class SyncDirsWatchdog(luigi.Task):
    src = luigi.Parameter()
    dst = luigi.Parameter()
    recursive = luigi.BoolParameter(default=True)
    patterns = luigi.Parameter(default="*")
    ignore_patterns = luigi.Parameter(default="")
    ignore_directories = luigi.BoolParameter(default=False)
    case_sensitive = luigi.BoolParameter(default=True)

    def run(self):
        from saisoku import WatchDog
        
        WatchDog(
            src=self.src, dst=self.dst, recursive=self.recursive, patterns=self.patterns,
            ignore_patterns=self.ignore_patterns, ignore_directories=self.ignore_directories, 
            case_sensitive=self.case_sensitive)
    

if __name__ == '__main__':
    luigi.run()


saisoku.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""saisoku.py
Saisoku is a Python module that helps you build complex pipelines of batch file copying jobs.

See README.md or https://github.com/shirosaidev/saisoku
for more information.

Author: shirosai <cpark16@gmail.com>

Copyright (C) Chris Park 2019
saisoku is released under the Apache 2.0 license. See
LICENSE for the full license text.
"""

import errno
import os
import sys
import time
import mmap
import shutil
import logging
import tempfile
try:
    from queue import Queue
except ImportError:
    from Queue import Queue
from threading import Thread, Lock
from shutil import copyfile, move, rmtree, Error
from scandir import scandir, walk
from tqdm import tqdm
from watchdog.observers import Observer
from watchdog.events import PatternMatchingEventHandler


SAISOKU_VERSION = '0.1-b.5'
__version__ = SAISOKU_VERSION


# settings
logtofile = True
loglevel = logging.DEBUG


fileQueue = Queue()


def logging_setup():
    """Set up logging."""
    logger = logging.getLogger(name='saisoku')
    logger.setLevel(loglevel)
    logformatter = logging.Formatter('%(asctime)s [%(levelname)s][%(name)s] %(message)s')
    ch = logging.StreamHandler()
    ch.setLevel(loglevel)
    ch.setFormatter(logformatter)
    logger.addHandler(ch)
    logger.propagate = False
    if logtofile:
        logfile = os.path.join(tempfile.gettempdir(), 'saisoku.log')
        hdlr = logging.FileHandler(logfile)
        hdlr.setLevel(loglevel)
        hdlr.setFormatter(logformatter)
        logger.addHandler(hdlr)
    return logger


logger = logging_setup()


class color:
    PURPLE = '\033[95m'
    CYAN = '\033[96m'
    DARKCYAN = '\033[36m'
    BLUE = '\033[94m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'
    END = '\033[0m'


def output_banner():
    import random
    c = random.choice((color.PURPLE, color.CYAN, color.YELLOW, color.RED))
    banner = '''%s

       ___              _                      _             
      / __|   __ _     (_)     ___     ___    | |__   _  _   
      \\__ \\  / _` |    | |    (_-<    / _ \\   | / /  | +| |  
      |___/  \\__,_|   _|_|_   /__/_   \___/   |_\\_\\   \\_,_|  
    _|"""""|_|"""""|_|"""""|_|"""""|_|"""""|_|"""""|_|"""""| 
    "`-0-0-'"`-0-0-'"`-0-0-'"`-0-0-'"`-0-0-'"`-0-0-'"`-0-0-' v%s

    %s''' % (c, SAISOKU_VERSION, color.END)
    print(banner)


def get_num_lines(fileNameList):
    """Get number of lines in txt file."""
    fp = open(fileNameList, "r+")
    buf = mmap.mmap(fp.fileno(), 0)
    lines = 0
    while buf.readline():
        lines += 1
    return lines


class ThreadedCopy:
    """Threaded file copy class."""
    totalFiles = 0
    copyCount = 0
    lock = Lock()

    def __init__(self, src, dst, threads=16, filelist=None, symlinks=False, ignore=None, copymeta=True, package=False):
        self.src = src
        self.dst = dst
        self.threads = threads
        self.filelist = filelist
        self.symlinks = symlinks
        # copytree ignore patterns like '*.pyc', 'tmp*'
        self.ignore = None if ignore is None else shutil.ignore_patterns(ignore)
        self.copymeta = copymeta
        self.is_package_task = package
        self.fileList = []
        self.sizecounter = 0
        self.errors = []

        logger.info('Starting file copy from %s to %s..' % (self.src, self.dst))

        # open filelist txt file or scandir src path and preprocess the total files sizes
        logger.info("Calculating total file size..")
        if filelist:
            with open(self.filelist, "r") as file:  # txt with a file per line
                for line in tqdm(file, total=get_num_lines(self.filelist), unit='files'):
                    fname = line.rstrip('\n')
                    if not self.is_package_task:  # copy files package task
                        fpath = os.path.join(self.src, fname)
                    else:
                        fpath = fname
                    size = os.stat(fpath).st_size
                    self.fileList.append((fname, fpath, size))
                    self.sizecounter += size
        else:
            for item in tqdm(scandir(self.src), unit='files'):  # scandir and build file list
                self.fileList.append(item)
                self.sizecounter += item.stat().st_size
        # make dst directory if it doesn't exist and copy stat
        try:
            os.makedirs(dst)
        except OSError as e:
            if e.errno == errno.EEXIST and os.path.isdir(dst):
                pass
            else:
                raise
        try:
            shutil.copystat(src, dst)
        except OSError as e:
            self.errors.extend((self.src, self.dst, str(e)))
    
        self.totalFiles = len(self.fileList)
        logger.info("Copying " + str(self.totalFiles) + " files (" + str(self.sizecounter) + " bytes)..")
        self.pbar = tqdm(total=self.sizecounter, unit='B', unit_scale=True, unit_divisor=1024)
        self.threadWorkerCopy(self.fileList)


    def CopyWorker(self):
        """Thread worker for file copying."""
        while True:
            fileName = fileQueue.get()
            try:
                isdir = fileName.is_dir()  # scandir object
                issym = fileName.is_symlink()
                size = fileName.stat().st_size
                fname = fileName.name
            except AttributeError:  # file from txt
                fname, fpath, size = fileName
                isdir = True if os.path.isdir(fpath) else False
                issym = True if os.path.islink(fpath) else False
            if not self.is_package_task:
                srcname = os.path.join(self.src, fname)
            else:
                srcname = fname
            dstname = os.path.join(self.dst, fname)
            try:
                if isdir:
                    #copyf = shutil.copy2 if self.copymeta else shutil.copyfile
                    #shutil.copytree(srcname, dstname, symlinks=self.symlinks, 
                    #                ignore=self.ignore, copy_function=copyf)
                    pass
                else:
                    if issym is self.symlinks:
                        if self.copymeta:
                            try:
                                shutil.copy2(srcname, dstname)
                            except (OSError, IOError) as e:
                                self.errors.extend((self.src, self.dst, str(e)))
                        else:
                            shutil.copyfile(srcname, dstname)
            except (OSError, IOError) as e:
                self.errors.append((srcname, dstname, str(e)))
            # catch the Error from the recursive copytree so that we can
            # continue with other files
            except Error as e:
                self.errors.extend(e.args[0])
            if self.errors:
                raise Error(self.errors)
            with self.lock:
                self.copyCount += 1
                #percent = (self.copyCount * 100) / self.totalFiles
                #print(str(percent) + " percent copied.")
                self.pbar.set_postfix(file=fname[-10:], refresh=False)
                self.pbar.update(size)
            fileQueue.task_done()

    def threadWorkerCopy(self, fileNameList):
        for i in range(self.threads):
            t = Thread(target=self.CopyWorker)
            t.daemon = True
            t.start()
        for fileName in fileNameList:
            fileQueue.put(fileName)
        fileQueue.join()
        self.pbar.close()
        logger.info('Done')


class ThreadedHTTPCopy:
    """Threaded HTTP file copy class."""
    totalFiles = 0
    copyCount = 0
    lock = Lock()

    def __init__(self, src, dst, threads=1, ports=[5005], fetchmode='urlretrieve', chunksize=8192):
        self.src = src
        self.dst = dst
        self.threads = threads
        self.ports = ports
        self.fileList = []
        self.sizecounter = 0
        self.fetchmode = fetchmode  # requests, urlretrieve
        self.chunksize = chunksize
        self.errors = []

        logger.info('Starting file copy from %s to %s..' % (self.src, self.dst))

        # get file list from http server
        logger.info("Getting file list from http server..")
        for item in tqdm(self.GetFileLinks(), unit='files'):  # get file links and build file list
            self.fileList.append(item)
            self.sizecounter += item[1]
        # make dst directory if it doesn't exist
        try:
            os.makedirs(dst)
        except OSError as e:
            if e.errno == errno.EEXIST and os.path.isdir(dst):
                pass
            else:
                raise
    
        self.totalFiles = len(self.fileList)
        logger.info("Copying " + str(self.totalFiles) + " files (" + str(self.sizecounter) + " bytes)..")
        self.pbar = tqdm(total=self.sizecounter, unit='B', unit_scale=True, unit_divisor=1024)
        self.threadWorkerCopy(self.fileList)


    def GetFileLinks(self):
        """Generator that yields tuple of file links and their size at url."""
        from bs4 import BeautifulSoup
        import requests

        url = self.tserv_lb()
        r = requests.get(url)
        html = r.content
        soup = BeautifulSoup(html, 'html.parser')
        for link in soup.find_all('a'):
            if not link.get('href').endswith('/'):  # files only
                yield (link.get('href'), int(link.get('title')))


    def FetchFile(self, src, dst):
        """Use urllib urlretrieve to fetch file."""
        try:
            from urllib import urlretrieve
        except ImportError:
            from urllib.request import urlretrieve
        import requests

        if self.fetchmode == 'urlretrieve':
            try:
                urlretrieve(src, dst)
            except Exception as e:
                self.errors.append((src, dst, str(e)))
        elif self.fetchmode == 'requests' or self.fetchmode is None:
            try:
                response = requests.get(src, stream=True)
                handle = open(dst, "wb")
                for chunk in response.iter_content(chunk_size=self.chunksize):
                    if chunk:
                        handle.write(chunk)
                handle.close()
            except Exception as e:
                self.errors.append((src, dst, str(e)))


    def tserv_lb(self):
        """Load balance across tserve ports."""
        import random
        port = random.choice(self.ports)
        url = self.src + ":" + str(port)
        return url


    def CopyWorker(self):
        """Thread worker for file copying."""
        try:
            from urlparse import urljoin
        except ImportError:
            from urllib.parse import urljoin

        while True:
            fileItem = fileQueue.get()
            fileName, size = fileItem
            url = self.tserv_lb()
            srcname = urljoin(url, fileName)
            dstname = os.path.join(self.dst, fileName)
            self.FetchFile(srcname, dstname)
            if self.errors:
                raise Error(self.errors)
            with self.lock:
                self.copyCount += 1
                self.pbar.set_postfix(file=fileName[-10:], refresh=False)
                self.pbar.update(size)
            fileQueue.task_done()


    def threadWorkerCopy(self, fileItemList):
        for i in range(self.threads):
            t = Thread(target=self.CopyWorker)
            t.daemon = True
            t.start()
        for fileItem in fileItemList:
            fileQueue.put(fileItem)
        fileQueue.join()
        self.pbar.close()
        logger.info('Done')


class Rclone:
    """Rclone class. Uses subprocess to run rclone."""

    def __init__(self, src, dst, flags=[], command='sync', cmdargs=[]):
        self.src = src
        self.dst = dst
        self.flags = flags
        self.command = command
        self.cmdargs = cmdargs
        self.errors = []

        self.run_rclone()

    def run_rclone(self):
        from subprocess import check_output, CalledProcessError, STDOUT

        cmd = ['rclone']
        [cmd.append(f) for f in self.flags]
        cmd.append(self.command)
        cmd.append(self.src)
        cmd.append(self.dst)
        [cmd.append(a) for a in self.cmdargs]
        logger.debug('rclone command: {}'.format(" ".join(cmd)))
        
        logger.info('Starting rclone from %s to %s..' % (self.src, self.dst))
        try:
            output = check_output(cmd, stderr=STDOUT)
            logger.debug(output)
        except CalledProcessError as e:
            self.errors.append((self.src, self.dst, str(e.output), str(e.returncode)))
        if self.errors:
            raise Error(self.errors)

        logger.info('Done')


class WatchDog:
    """WatchDog class. Uses watchdog python module."""

    def __init__(self, src, dst, recursive, patterns, ignore_patterns, ignore_directories, case_sensitive):
        self.src = src
        self.dst = dst
        self.recursive = recursive
        self.patterns = patterns
        self.ignore_patterns = ignore_patterns
        self.ignore_directories = ignore_directories
        self.case_sensitive = case_sensitive
        self.errors = []

        self.sync_dir()

    def sync_dir(self):
        logger.info(f'Doing initial sync from {self.src} to {self.dst}...')
        os.system('rsync -av "{0}" "{1}"'.format(self.src+'/', self.dst+'/'))
        logger.info(f'Starting watchdog sync from {self.src} to {self.dst}')
        event_handler = PatternMatchingEventHandler(
            self.patterns, self.ignore_patterns, self.ignore_directories, self.case_sensitive)
        event_handler.on_created = self.on_created
        event_handler.on_deleted = self.on_deleted
        event_handler.on_modified = self.on_modified
        event_handler.on_moved = self.on_moved
        observer = Observer()
        observer.schedule(event_handler, self.src, recursive=self.recursive)
        observer.start()
        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            observer.stop()
        observer.join()
        logger.info('Watchdog sync stopped')

    def on_created(self, event):
        logger.info(f"{event.src_path} has been created")
        if not event.is_directory:
            try:
                file_size = -1
                while file_size != os.path.getsize(event.src_path):
                    file_size = os.path.getsize(event.src_path)
                    time.sleep(1)
                copyfile(event.src_path, os.path.join(self.dst, event.src_path.replace(self.src, self.dst)))
            except FileExistsError:
                pass

    def on_deleted(self, event):
        logger.info(f"{event.src_path} has been deleted")
        try:
            logger.info(f"deleting {os.path.join(self.dst, event.src_path.replace(self.src, self.dst))}")
            if not event.is_directory:
                os.remove(os.path.join(self.dst, event.src_path.replace(self.src, self.dst)))
            else:
                rmtree(os.path.join(self.dst, event.src_path.replace(self.src, self.dst)), ignore_errors=True)
        except FileNotFoundError:
            pass

    def on_modified(self, event):
        logger.info(f"{event.src_path} has been modified")
        if not event.is_directory:
            os.system(f'rsync -uv "{event.src_path}" "{os.path.join(self.dst, event.src_path.replace(self.src, self.dst))}"')
            
    def on_moved(self, event):
        logger.info(f"{event.src_path} moved to {event.dest_path}")
        if event.is_directory:
            for dirpath, subdirs, files in walk(self.dst):
                if dirpath == event.src_path.replace(self.src, self.dst):
                    move(dirpath, event.dest_path.replace(self.src, self.dst))
        else:
            for dirpath, subdirs, files in walk(self.dst):
                for file in files:
                    filepath = os.path.join(dirpath, file)
                    if filepath == event.src_path.replace(self.src, self.dst):
                        move(filepath, event.dest_path.replace(self.src, self.dst))
